lib/Runtime/Executor/ThreadPoolExecutor.cpp

```diff
@@ -351,6 +389,14 @@ void ThreadPoolExecutor::handleDeviceManagerResult(
     std::lock_guard<std::mutex> lock(executionStatesMutex_);
     executionStates_.erase(executionState->getRunId());
```

Comment:
might be missing something, but should we unblock in dtor of ThreadPoolExecutor once executionState_ is empty?
In this case we could simply signal on some CV if executionStates_ is empty and do a wait in dtor. In this case change would be less intrusive (no need for additional primitives, etc).